### DISCUSSION TITLE
3D classification tutorial

### DISCUSSION THREAD

**Discussion by hermitico** [Project-MONAI/tutorials#929] [Q&A]
Dear all,

I'm new to this field, and I find your project remarkable.
I would like to know if it's possible to turn that classification example into a regression.
Instead of predicting sex based on structural MRI, I would like to predict cognitive scores having continuous variables.
If so, what lines of code would you add/change to the tutorial example code?

Thanks

**Comment by KumoLiu** [comment]
Hi @hermitico, thank you for your question. If you want to run a regression task, you can simply follow these steps:
1. Remove the activation function like this
https://github.com/Project-MONAI/tutorials/blob/fa7c06aedc87416154beb821ca15816722a1ad91/3d_classification/torch/densenet_training_dict.py#L78
2. Change the loss function to `MSELOSS`.
Hope it can help you, thanks!

**Comment by hermitico** [comment]
Thanks! 

I have commented post_pred. What about post_label?
#post_pred = Compose([Activations(softmax=True)])
post_label = Compose([AsDiscrete(to_onehot=2)])

 **y_onehot** and  **y_pred_act** are prepared for classification instead of regression, right?
The transformation is post_pred is throwing an error:
applying transform <monai.transforms.post.array.AsDiscrete object at 0x7f257b6f3890>

How should I change this validation part?

    if (epoch + 1) % val_interval == 0:
        model.eval()
        with torch.no_grad():
            y_pred = torch.tensor([], dtype=torch.float32, device=device)
            y = torch.tensor([], dtype=torch.long, device=device)
            for val_data in val_loader:
                val_images, val_labels = val_data["img"].to(device), val_data["label"].to(device)
                y_pred = torch.cat([y_pred, model(val_images)], dim=0)
                y = torch.cat([y, val_labels], dim=0)

            acc_value = torch.eq(y_pred.argmax(dim=1), y)
            acc_metric = acc_value.sum().item() / len(acc_value)
            **y_onehot = [post_label(i) for i in decollate_batch(y, detach=False)]**
            **y_pred_act = [post_pred(i) for i in decollate_batch(y_pred)]**
            auc_metric(y_pred_act, y_onehot)
            auc_result = auc_metric.aggregate()
            auc_metric.reset()
            del y_pred_act, y_onehot
            if acc_metric > best_metric:
                best_metric = acc_metric
                best_metric_epoch = epoch + 1
                torch.save(model.state_dict(), "best_metric_model_classification3d_dict.pth")
                print("saved new best metric model")
            print(
                "current epoch: {} current accuracy: {:.4f} current AUC: {:.4f} best accuracy: {:.4f} at epoch {}".format(
                    epoch + 1, acc_metric, auc_result, best_metric, best_metric_epoch
                )
            )
            writer.add_scalar("val_accuracy", acc_metric, epoch + 1)
            print(f"train completed, best_metric: {best_metric:.4f} at epoch: {best_metric_epoch}")
            writer.close()

**Comment by KumoLiu** [comment]
Hi @hermitico, you also don't need to discrete your label. And the output of the network is your prediction, which is used to calculate the loss and metric with your label. So in validation part, you also don't need prepare your output and label, use `y` and `y_pred` to calculate the loss and metric.
```
y_onehot = [post_label(i) for i in decollate_batch(y, detach=False)]
y_pred_act = [post_pred(i) for i in decollate_batch(y_pred)]
```
Hope it can help you! Thanks!

**Reply by marafathussain** [reply]
@KumoLiu Hi, I have a related question if you don't mind suggesting some solution. I need to use GradCAM (or any other CAM version from MONAI visualization) on regression problems. I have a single regressed output value. But MONAI GradCAM seems to work on classification problems only. Thanks in advance.

**Reply by KumoLiu** [reply]
Hi @marafathussain, you can still visualize this with a visualization tool such as GradCAM. The only difference is that the output of the regression task has no concept of classes, and the activation plot you see does not indicate the probability of a class is activated. Hope it can help you. Thanks!
